### PR TITLE
Fixed support to contact both ipv4 and ipv6 hosts

### PR DESCRIPTION
**Pull request by QuantumEntangledAndy**
This should allow one to contact both ipv4 and ipv6 (provided listen_on is set to an ipv6 address). 

There were two issues one resolving host names as described in #196 and two the ipv4 addresses needed to be masked to ipv6 address on macos (ubuntu seemed to do this implicitly).

The first issue was resolved by specializing the resolve template to handle explicit ip addresses differently to the protocol in the settings.

The second issue was resolved by masking ipv4 addresses if the socket is using the ipv6 protocol. There was already code in the library to do this but it seems that it was left uncalled by some prior commit. Perhaps you know when the app stopped masking the ipv4 addresses when needed? Let me know if that causes some edge cases that I might need to work around.

Fixes #197

### CODE REVIEW

**Comment by QuantumEntangledAndy** [comment]
Travis seems to be failing just on the linux build like before. I might make that my next pull request.

**Comment by QuantumEntangledAndy** [comment]
Don't pull just yet I am trying a different method to resolve the ipv4 addresses on ipv6 protocol by masking them first. It is a bit cleaner than the method I am using here.

**Comment by QuantumEntangledAndy** [comment]
Ok that should work well. I still wonder if it would be better to not try to resolve ip addresses but just call the handeler without going through the boost resolver.

**Comment by QuantumEntangledAndy** [comment]
Ok so it turns out there was a simpler solution, than my original attempt. If you don't specify a protocol then it is possible to resolve ipv4 and ipv6. Also there's a flag in boost asio boost::asio::ip::resolver_query_base::numeric_host that will make it not try to resolve the address and just treat is as a ip as is.

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp [hunk 1]
```diff
@@ -226,9 +226,12 @@ namespace asiotap
 	template <typename AddressType>
 	inline void async_resolve(const base_ip_endpoint<AddressType>& ep, typename base_ip_endpoint<AddressType>::resolver& resolver, typename base_ip_endpoint<AddressType>::resolver::protocol_type protocol, typename base_ip_endpoint<AddressType>::resolver::query::flags flags, const std::string& default_service, typename base_ip_endpoint<AddressType>::handler handler)
 	{
-		typename base_ip_endpoint<AddressType>::resolver::query query(protocol, ep.address().to_string(), ep.has_port() ? boost::lexical_cast<std::string>(ep.port()) : default_service, flags);
-
+		static_cast<void>(flags);
+		static_cast<void>(protocol);
+		typename base_ip_endpoint<AddressType>::resolver::query query(ep.address().to_string(), ep.has_port() ? boost::lexical_cast<std::string>(ep.port()) : default_service, boost::asio::ip::resolver_query_base::numeric_host);
+
```
Comment (s-vincent) [inline libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: I would keep the flags parameter in the call : flags | boost::asio::ip::resolver_query_base::numeric_host

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp [hunk 2]
```diff
@@ -226,9 +226,12 @@ namespace asiotap
...
 	inline void async_resolve(const base_ip_endpoint<AddressType>& ep, typename base_ip_endpoint<AddressType>::resolver& resolver, typename base_ip_endpoint<AddressType>::resolver::protocol_type protocol, typename base_ip_endpoint<AddressType>::resolver::query::flags flags, const std::string& default_service, typename base_ip_endpoint<AddressType>::handler handler)
 	{
-		typename base_ip_endpoint<AddressType>::resolver::query query(protocol, ep.address().to_string(), ep.has_port() ? boost::lexical_cast<std::string>(ep.port()) : default_service, flags);
-
+		static_cast<void>(flags);
+		static_cast<void>(protocol);
+		typename base_ip_endpoint<AddressType>::resolver::query query(ep.address().to_string(), ep.has_port() ? boost::lexical_cast<std::string>(ep.port()) : default_service, boost::asio::ip::resolver_query_base::numeric_host);
+		
 		resolver.async_resolve(query, handler);
+		return;
```
Comment (s-vincent) [inline libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: "return" is not needed here.

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp [hunk 3]
```diff
@@ -226,9 +226,12 @@ namespace asiotap
 	template <typename AddressType>
 	inline void async_resolve(const base_ip_endpoint<AddressType>& ep, typename base_ip_endpoint<AddressType>::resolver& resolver, typename base_ip_endpoint<AddressType>::resolver::protocol_type protocol, typename base_ip_endpoint<AddressType>::resolver::query::flags flags, const std::string& default_service, typename base_ip_endpoint<AddressType>::handler handler)
 	{
-		typename base_ip_endpoint<AddressType>::resolver::query query(protocol, ep.address().to_string(), ep.has_port() ? boost::lexical_cast<std::string>(ep.port()) : default_service, flags);
-
+		static_cast<void>(flags);
```
Comment (s-vincent) [inline libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: To remove, see next comment to add flags in the function call.

**Comment by s-vincent** [comment]
Thank you for the PR !

LGTM except very minor changes.

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp (hunk 2 above)
Comment (QuantumEntangledAndy) [reply libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: Ok removed :)

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp (hunk 3 above)
Comment (QuantumEntangledAndy) [reply libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: static_cast<void>(flags); has been removed as its now as used paramter

File: libs/asiotap/include/asiotap/types/ip_endpoint.hpp (hunk 1 above)
Comment (QuantumEntangledAndy) [reply libs/asiotap/include/asiotap/types/ip_endpoint.hpp]: Flags are now combined with new commit

**Comment by QuantumEntangledAndy** [comment]
Shall I squash the commits into one?

**Comment by s-vincent** [comment]
Thanks !

**Pull request merged**